### ISSUE TITLE
Should HttpServerCodec issue 100 continue responses when there's no HttpObjectAggregator in the pipeline?

### ISSUE THREAD

**Issue by dschobel** [netty/netty#5180]
A http server pipeline with both HttpServerCodec and HttpObjectAggregator installed will do the right thing for requests with 'expect: 100 continue' headers but removing the object aggregator leaves users rolling their own handling as seen in [netty's http server hello world example]( https://github.com/netty/netty/blob/4.1/example/src/main/java/io/netty/example/http/helloworld/HttpHelloWorldServerHandler.java#L48).

If HttpServerCodec is intended to be the 'batteries included' protocol implementation it seems like it could take the burden off application writers by issuing 100 CONTINUE responses when there is no HttpObjectAggregator instance in the pipeline thereby saving clients the timeout before proceeding with the payload.

This behavior exists in 4.1.0.CR7 (and has for a long time I think).

**Comment by rkapsi** [comment]
It depends on the use-case and would have to be a configurable option. We're using Netty to implement something like a proxy server and it'd be fatal if HttpServerCodec would start issuing 100s on its own.

**Comment by Scottmitch** [comment]
@dschobel - The `HttpObjectAggregator` is designed to give users a "simple" `FullHttpMessage` in the pipeline, and has some knowledge of the `maxRequestSize`. If `HttpObjectAggregator` can infer the body size from the headers it can make an informed decision and accept/reject the continue request, and if it cannot infer the body size it will enforce `maxRequestSize` as data is read. `HttpServerCodec` currently has no knowledge of how much data is OK, or other constraints that may impact the server's response to a continue request. If you are interested in submitting a PR to make `HttpServerCodec` accept some sort of pluggable policy (as indicated by @rkapsi) can you elaborate on what ideas you have in this area?

**Comment by dschobel** [comment]
@Scottmitch yep, a `HttpServerCodec` application without `HttpObjectAggregator` doesn't have enough information to intelligently reject/accept based on headers but by ignoring the header we're not changing the outcome for clients, when netty is acting as an origin server we're only delaying the request which leaves application writers having to include some boiler plate as in the hello world example.

As @rkapsi points out, ignoring the header does make `HttpServerCodec` usable for proxying but that's the crux of the issue in my view, it seems like arbitrary coupling to tie `HttpObjectAggregator` with origin server friendly behavior (managing 100 CONTINUE responses) and `HttpServerCodec` to proxy friendly behavior.

There are a lot of ways to skin the configuration cat, we could either say that `HttpServerCodec` and `HttpObjectAggregator` both will do the origin server friendly thing of managing 100 CONTINUE and then people who want to build proxies can either use `HttpResponseDecoder` directly, or we could introduce some parallel HttpProxyServerCodec handler which ignore the header, or otherwise have a policy param on `HttpServerCodec`.

**Comment by Scottmitch** [comment]
> we could either say that HttpServerCodec and HttpObjectAggregator both will do the origin server friendly thing of managing 100 CONTINUE

It seems a bit strange to have the 100 continue maintained in 2 places. `HttpServerCodec` and `HttpObjectAggregator` can exist in the pipeline at the same time so it may be a confusing to have the logic duplicated in the pipeline.

> it seems like arbitrary coupling to tie `HttpObjectAggregator` with origin server friendly behavior

IIUC the intention is not focused on being origin/proxy friendly but more that if `HttpObjectAggregator` didn't handle 100-continue the server may not make progress. Users that put `HttpObjectAggregator` in the pipeline only expect to handle `FullHttpMessage` objects after this, and if `HttpObjectAggregator` doesn't respond to the 100-continue the client may never issue the request.

> or otherwise have a policy param on HttpServerCodec.

I would be interested in hearing more here. My concerns are duplication of concerns and maintaining existing behavior by default.

**Comment by dschobel** [comment]
> It seems a bit strange to have the 100 continue maintained in 2 places. HttpServerCodec and HttpObjectAggregator can exist in the pipeline at the same time so it may be a confusing to have the logic duplicated in the pipeline.


That's fair but I think tolerable since it will be a (small I think) burden on the codec maintainers vs the benefit to the server writers of not timing out their clients expect-continue timer or copying boiler plate.

I think the complexity will amount to:

```
if (!object aggregator in pipeline && request has expect: continue header)
  send 100 CONTINUE
```

and the object aggregator doesn't need to be touched.




If the goal is to be tolerant of dodgey clients that wait forever for the 100 CONTINUE then it seems reasonable to be tolerant even when the server isn't aggregating chunks.


> I would be interested in hearing more here. My concerns are duplication of responsibility and maintaining existing behavior by default.


Concretely, I think this can solved in a small way with the change above to HttpServerCodec. If we want to be behavior compatible with the current implementation, we can paramterize the ctor and default behavior off.
Even then the expect-continue handling becomes more visible to users as it becomes part of the java doc rather than a snippet of code to copy/paste from the tutorial.

If you're open to a big change (ie; for the 5.x branch) then there can be a single overarching HttpServerCodec which manages the codec + aggregator handlers for you and will save us from having to introspect on the pipeline. Whether to aggregate chunks becomes a param on the new super-codec and expect-continue behavior is a policy param.

**Comment by Scottmitch** [comment]
@dschobel - It is not ideal for one handler to look for another handler's presence in the pipeline ... kinda breaks the boundaries between handlers. What about creating some sort of `HttpContinueRequestHandler` which can intercept the "expect continue" request and call some overridable method which by default just responds with 100 continue. This way the user would have to explicitly add this into the pipeline to get this behavior. So the pipeline would look like:

```java
pipeline.addLast(new HttpServerCodec(...));
pipeline.addLast(new HttpContinueRequestHandler(...)); // by default just send 100-continue
// It is now much more clear that the previous handler will handle continue requests and does
// not change the existing behavior of any handlers ...
```

**Comment by dschobel** [comment]
@Scottmitch that definitely works, it's the solution we arrived at for finagle where the handler is conditionally installed depending on whether chunk aggregation is configured.

```
pipeline.addLast(HttpServerCodec(...))
if (streaming) { pipeline.addLast(new HttpContinueRequestHandler(...)) } 
else { pipeline.addLast(HttpObjectAggregator(...)) } 
```

that's already less boilerplate so I'm happy with that if inspecting the pipeline is a non-starter and bundling aggregation as part of a new handler that manages the codec is too big of a change.

**Comment by Scottmitch** [comment]
@dschobel - Yes this is what I had in mind ... looks like you guys already figured it out :)  Are you interested in contributing this to Netty?

> if inspecting the pipeline is a non-starter

best to avoid this if we can IMO

**Comment by dschobel** [comment]
sounds good, I'll take a crack at the PR next week.

**Comment by normanmaurer** [comment]
@dschobel thanks buddy!

**Comment by normanmaurer** [comment]
@dschobel ping

**Comment by dschobel** [comment]
@normanmaurer I'm going on leave for 6 weeks as of today so I won't get to this before I'm back but I'd still like to contribute this back from finagle.

**Comment by normanmaurer** [comment]
@dschobel ping
